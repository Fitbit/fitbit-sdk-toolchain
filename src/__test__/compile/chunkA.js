import('./chunkB')
    .then(({ chunkB }) => {
        chunkB();
    })
