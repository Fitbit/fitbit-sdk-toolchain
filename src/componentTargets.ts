interface ComponentTarget {
  inputs: string[];
  outputDir?: string;
  notFoundIsFatal: boolean;
  translationsGlob: string;
}

export enum ComponentType {
  DEVICE = 'device',
  COMPANION = 'companion',
  SETTINGS = 'settings',
}

const componentTarget: { [component in ComponentType]: ComponentTarget } = {
  [ComponentType.DEVICE]: {
    inputs: ['app/index.ts', 'app/index.js'],
    outputDir: 'app',
    notFoundIsFatal: true,
    translationsGlob: 'app/i18n/*.po',
  },
  [ComponentType.COMPANION]: {
    inputs: ['companion/index.ts', 'companion/index.js'],
    notFoundIsFatal: false,
    translationsGlob: 'companion/i18n/*.po',
  },
  [ComponentType.SETTINGS]: {
    inputs: ['tsx', 'ts', 'jsx', 'js'].map((ext) => `settings/index.${ext}`),
    notFoundIsFatal: false,
    translationsGlob: 'settings/i18n/*.po',
  },
};

export default componentTarget;
