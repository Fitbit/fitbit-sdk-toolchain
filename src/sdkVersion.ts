import semver from 'semver';

import packageVersionConst from './packageVersion.const';

export default function sdkVersion(toolchainVersion = packageVersionConst) {
  const version = semver.parse(toolchainVersion);
  if (version === null) {
    throw new Error(`Invalid SDK package version: ${toolchainVersion}`);
  }
  return version;
}

export function apiVersions(
  { enableProposedAPI }: { enableProposedAPI?: boolean } = {},
  toolchainVersion = packageVersionConst,
) {
  if (enableProposedAPI) return { deviceApi: '*', companionApi: '*' };

  const { major, minor } = sdkVersion(toolchainVersion);
  if (major === 1 && minor === 0) {
    return { deviceApi: '1.0.0', companionApi: '1.0.0' };
  }
  if (major === 2 && minor === 0) {
    return { deviceApi: '3.0.0', companionApi: '2.0.0' };
  }
  if (major === 3 && minor === 0) {
    return { deviceApi: '4.0.0', companionApi: '2.1.0' };
  }
  throw new Error(
    `No known API versions for SDK package version ${major}.${minor}`,
  );
}
