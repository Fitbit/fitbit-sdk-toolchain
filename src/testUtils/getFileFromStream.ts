import stream from 'stream';

import Vinyl from 'vinyl';

export default function getFileFromStream(stream: stream.Readable, filename?: string): Promise<Vinyl> {
    return new Promise<Vinyl>((resolve, reject) => {
        stream.on('data', (file: Vinyl) => {
            if (file.contents && (!filename || file.basename === filename)) {
                resolve(file);
            }
        });
        stream.on('finish', () => reject(new Error('No file in stream matched')));
        stream.on('error', reject);
    });
}
