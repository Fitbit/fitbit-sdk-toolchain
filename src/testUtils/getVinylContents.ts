import concatStream from 'concat-stream';
import Vinyl from 'vinyl';

export default function getVinylContents(file: Vinyl) {
  return new Promise<string>((resolve, reject) => {
    if (file.isBuffer()) resolve(file.contents.toString());
    else if (file.isStream()) {
      file.contents.pipe(concatStream({ encoding: 'string' }, (str) => {
        // Typing for concatStream is broken, thinks it always returns buffers
        resolve(str as any);
      }));
    }
    else reject(new Error(`Unknown Vinyl file type for path: ${file.path}`));
  });
}
