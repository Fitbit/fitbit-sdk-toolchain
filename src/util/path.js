export function dirname(path) {
  let dir = path.split('/').slice(0, -1).join('/');
  if (dir === '') dir = '/';
  return dir;
}

export function basename(path) {
  return path.replace(/\\/g, '/').replace(/.*\//, '');
}

export function split(path) {
  return [dirname(path), basename(path)];
}
