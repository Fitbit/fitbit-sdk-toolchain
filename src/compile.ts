import path from 'path';

import * as rollup from 'rollup';
import rollupPluginBabel from 'rollup-plugin-babel';
import rollupPluginCommonjs from 'rollup-plugin-commonjs';
import rollupPluginNodeResolve from 'rollup-plugin-node-resolve';
import ts from 'typescript';

import componentTargets, { ComponentType } from './componentTargets';
import {
  DiagnosticCategory,
  DiagnosticHandler,
  logDiagnosticToConsole,
} from './diagnostics';
import rollupToVinyl from './rollupToVinyl';
import sdkVersion from './sdkVersion';

import forbidAbsoluteImport from './plugins/forbidAbsoluteImport';
import i18nPolyfill from './plugins/i18nPolyfill';
import platformExternals from './plugins/platformExternals';
import polyfill, { PolyfillMap } from './plugins/polyfill';
import polyfillDevice from './plugins/polyfillDevice';
import resourceImports from './plugins/resourceImports';
import terser from './plugins/terser';
import typescript from './plugins/typescript';
import rollupWarningHandler from './rollupWarningHandler';

// TODO: emit a warning when any of these settings are
// defined in the app's tsconfig
const tsconfigOverrides = {
  noEmitHelpers: false,
  importHelpers: true,
  noResolve: false,
  noEmit: false,
  inlineSourceMap: false,
  moduleResolution: ts.ModuleResolutionKind.NodeJs,
  module: ts.ModuleKind.ES2015,
  suppressOutputPathCheck: true,
};

function pluginIf(condition: boolean, plugin: () => rollup.Plugin) {
  return condition ? [plugin()] : [];
}

export default function compile({
  component,
  entryPoint,
  outputDir,
  defaultLanguage,
  allowUnknownExternals = false,
  onDiagnostic = logDiagnosticToConsole,
  polyfills = {},
}: {
  component: ComponentType;
  entryPoint: string;
  outputDir?: string;
  defaultLanguage: string;
  allowUnknownExternals?: boolean;
  onDiagnostic?: DiagnosticHandler;
  polyfills?: PolyfillMap;
}) {
  const ecma = component !== ComponentType.DEVICE ? 6 : 5;
  const { translationsGlob } = componentTargets[component];
  return rollupToVinyl(
    {
      input: entryPoint,
      plugins: [
        // Polyfills must come before platform externals
        polyfill(polyfills),
        ...pluginIf(
          (sdkVersion().major >= 4 ||
            (sdkVersion().major === 3 && sdkVersion().minor >= 1)) &&
            component === ComponentType.DEVICE,
          polyfillDevice,
        ),
        ...pluginIf(
          (sdkVersion().major >= 4 ||
            (sdkVersion().major === 3 && sdkVersion().minor >= 1)) &&
            component !== ComponentType.DEVICE,
          () => polyfill(i18nPolyfill(translationsGlob, defaultLanguage)),
        ),
        platformExternals(component),
        typescript({
          onDiagnostic,
          tsconfigOverride: {
            ...tsconfigOverrides,
            target: ecma === 6 ? ts.ScriptTarget.ES2015 : ts.ScriptTarget.ES5,
          },
          tsconfigSearchPath: path.dirname(entryPoint),
        }),
        ...pluginIf(component === ComponentType.SETTINGS, resourceImports),
        forbidAbsoluteImport(),
        rollupPluginNodeResolve({ preferBuiltins: false }),
        rollupPluginCommonjs({ include: ['node_modules/**'] }),
        ...pluginIf(ecma === 5, () =>
          rollupPluginBabel({
            plugins: [
              // Plugins are specified in this way to avoid this:
              // https://github.com/webpack/webpack/issues/1866
              // Also makes this work correctly in a browser environment
              require('@babel/plugin-transform-block-scoped-functions'),
              require('@babel/plugin-transform-block-scoping'),
              require('@babel/plugin-syntax-dynamic-import'),
            ],
            compact: false,
            babelrc: false,
            // We include JSON here to get a more sane error that includes the path
            extensions: ['.js', '.json'],
            // Types for babel are broken and don't accept anything but an object here
            inputSourceMap: false as any,
          }),
        ),
        terser({
          ecma,
          // We still support iOS 10, which ships Safari 10
          safari10: component !== ComponentType.DEVICE,
          mangle: {
            toplevel: true,
          },
          output: {
            // Fitbit OS versions before 2.2 couldn't handle multiple statements per line and still
            // give correct position info
            // Mobile doesn't give correct column info, so also one statement per line
            // Happily this causes a negligible difference in code size
            semicolons: false,
          },
          // Compression produces bad source maps
          // https://github.com/mishoo/UglifyJS2#source-maps-and-debugging
          compress: false,
        }),
      ],
      onwarn: rollupWarningHandler({
        onDiagnostic,
        codeCategories: allowUnknownExternals
          ? { UNRESOLVED_IMPORT: DiagnosticCategory.Warning }
          : undefined,
      }),
      inlineDynamicImports: true,
    },
    {
      dir: outputDir,
      file: outputDir === undefined ? `${component}.js` : undefined,
      format: 'cjs',
      sourcemap: true,
      // mapPath is relative to outputDir,
      sourcemapPathTransform: (mapPath) =>
        path.normalize(
          outputDir === undefined ? mapPath : path.join(outputDir, mapPath),
        ),
    },
  );
}
