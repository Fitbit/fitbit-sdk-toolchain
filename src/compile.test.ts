import path from 'path';

import { SourceMapConsumer } from 'source-map';
import ts from 'typescript';
import Vinyl from 'vinyl';

import compile from './compile';
import sdkVersion from './sdkVersion';
import getFileFromStream from './testUtils/getFileFromStream';
import getVinylContents from './testUtils/getVinylContents';
import { ComponentType } from './componentTargets';

jest.mock('./sdkVersion');

expect.addSnapshotSerializer({
  test: (val) => typeof val === 'string' && val.includes(process.cwd()),
  print: (val) => val.replace(process.cwd(), '<cwd>'),
});

let mockDiagnosticHandler: jest.Mock;
const mockSDKVersion = sdkVersion as jest.Mock;

beforeEach(() => {
  mockDiagnosticHandler = jest.fn();
  mockSDKVersion.mockReturnValue({ major: 2, minor: 0 });

  // We don't want to load the actual tsconfig.json for this project
  // during unit tests. Using a real tsconfig.json located within
  // the test assets folder speeds up unit tests too, by restricting
  // tsc to compiling within this directory.
  jest
    .spyOn(ts, 'findConfigFile')
    .mockReturnValue(testResourcePath('tsconfig.json'));
});

function testResourcePath(...paths: string[]) {
  return path.resolve(__dirname, '__test__', 'compile', ...paths);
}

function compileFile(
  filename: string,
  {
    allowUnknownExternals = undefined as boolean | undefined,
    component = ComponentType.COMPANION,
    outputDir = undefined as string | undefined,
    polyfills = {},
  } = {},
  expectFilename?: string,
) {
  return getFileFromStream(
    compile({
      component,
      allowUnknownExternals,
      outputDir,
      polyfills,
      entryPoint: testResourcePath(filename),
      onDiagnostic: mockDiagnosticHandler,
      defaultLanguage: 'en-US',
    }),
    expectFilename,
  );
}

// Build and assert correct output
it.each([
  ['building a minimal app', 'basic.js'],
  ['tslib when imported', 'providesTslib.js'],
  ['known external imports are used', 'knownExternalImport.js'],
  ['require calls are unmodified', 'require.js'],
  ['compiles a JS file with tslib helpers inline', 'inlineHelpers.js'],
  ['compiles a JS file with tslib helpers imported', 'importHelpers.js'],
])('build passes when %s', (_, filename) =>
  expect(
    compileFile(filename).then(getVinylContents),
  ).resolves.toMatchSnapshot(),
);

it('emits files in a specified output directory', () =>
  expect(
    compileFile(
      'basic.js',
      { outputDir: 'some_directory' },
      'some_directory/basic.js',
    ),
  ).resolves.toBeDefined());

// Just check build is successful
it.each([['importing a package', 'importPackage.js']])(
  'build passes when %s',
  (_, filename) =>
    expect(compileFile(filename).then(getVinylContents)).resolves.toBeDefined(),
);

it('allows importing image files when building settings', () =>
  expect(
    compileFile('importImage.js', { component: ComponentType.SETTINGS }).then(
      getVinylContents,
    ),
  ).resolves.toMatchSnapshot());

describe('when targeting SDK 1.0', () => {
  beforeEach(() => mockSDKVersion.mockReturnValue({ major: 1, minor: 0 }));

  it('allows JSON imports', () =>
    expect(
      compileFile('importJSON.js').then(getVinylContents),
    ).resolves.toMatchSnapshot());

  it('allows unintentionally non-relative imports', async () => {
    await expect(
      compileFile('incorrectlyNonRelativeImport.js').then(getVinylContents),
    ).resolves.toMatchSnapshot();
    expect(mockDiagnosticHandler.mock.calls[0]).toMatchSnapshot();
  });

  it.each([ComponentType.DEVICE, ComponentType.COMPANION])(
    'allows importing image files when building %s',
    (component: ComponentType) =>
      expect(
        compileFile('importImage.js', { component }),
      ).resolves.toBeDefined(),
  );
});

describe('when targeting SDK 2.0', () => {
  beforeEach(() => mockSDKVersion.mockReturnValue({ major: 2, minor: 0 }));

  it('does not allow JSON imports', () =>
    expect(
      compileFile('importJSON.js').then(getVinylContents),
    ).rejects.toThrowErrorMatchingSnapshot());

  it('does not allow unintentionally non-relative imports', () =>
    expect(
      compileFile('incorrectlyNonRelativeImport.js').then(getVinylContents),
    ).rejects.toThrowError());

  it.each([ComponentType.DEVICE, ComponentType.COMPANION])(
    'allows importing image files when building %s',
    (component: ComponentType) =>
      expect(
        compileFile('importImage.js', { component }),
      ).resolves.toBeDefined(),
  );
});

describe('when targeting SDK 3.0', () => {
  beforeEach(() => mockSDKVersion.mockReturnValue({ major: 3, minor: 0 }));

  it('emits ES6 code', () =>
    expect(
      compileFile('ES6.js').then(getVinylContents),
    ).resolves.toMatchSnapshot());

  it('emits ES5 code for device', () =>
    expect(
      compileFile('ES6.js', { component: ComponentType.DEVICE }).then(
        getVinylContents,
      ),
    ).resolves.toMatchSnapshot());

  it.each([ComponentType.DEVICE, ComponentType.COMPANION])(
    'does not allow importing image files when building %s',
    (component: ComponentType) =>
      expect(
        compileFile('importImage.js', { component }),
      ).rejects.toBeDefined(),
  );
});

it.each([
  ['an unrecognized binary file is imported', 'importBinary.js'],
  ['a non-existent relative import is specified', 'relativeImportNotFound.js'],
  ['unknown external imports are used', 'unknownExternalImport.js'],
  ['an absolute import is specified', 'absoluteImport.js'],
])('build fails when %s', (_, filename) =>
  expect(compileFile(filename)).rejects.toThrowErrorMatchingSnapshot(),
);

it.each([
  ['JS code with bad syntax', 'badSyntax.js'],
  ['TS code with a type error', 'typeError.ts'],
])('emits diagnostics given %s', async (_, filename) => {
  await expect(compileFile(filename)).rejects.toThrowError();
  expect(mockDiagnosticHandler.mock.calls[0]).toMatchSnapshot();
});

describe('when compiling a module with statements above an import declaration', () => {
  let buildOutput: Promise<Vinyl>;

  beforeEach(() => {
    buildOutput = compileFile('statementsBeforeImport.js');
  });

  it('builds', () =>
    expect(buildOutput.then(getVinylContents)).resolves.toMatchSnapshot());

  // Test for a regression (IPD-78696) that caused invalid sourcemaps to be produced
  it('emits a sourcemap with the first statement after the import correctly mapped', async () => {
    const file = await buildOutput;
    const code = await getVinylContents(file);

    const bundleLines = code.split('\n');
    const marker = 'console.log';
    const logLine = bundleLines.findIndex((line) => line.includes(marker));
    const logColumn = bundleLines[logLine].indexOf(marker);

    const map = await new SourceMapConsumer(file.sourceMap);
    expect(
      map.originalPositionFor({ line: logLine + 1, column: logColumn }),
    ).toMatchObject({ line: 3, column: 0 });
  });
});

describe('when allowUnknownExternals is enabled', () => {
  it('successfully builds an app with unknown external imports', async () => {
    await expect(
      compileFile('unknownExternalImport.js', {
        allowUnknownExternals: true,
      }).then(getVinylContents),
    ).resolves.toMatchSnapshot();
    expect(mockDiagnosticHandler.mock.calls).toMatchSnapshot();
  });
});

describe('when building a device component which uses gettext', () => {
  let file: string;

  beforeEach(async () => {
    mockSDKVersion.mockReturnValue({ major: 3, minor: 0 });
    file = await compileFile('i18n.js', {
      component: ComponentType.DEVICE,
    }).then(getVinylContents);
  });

  it('polyfills gettext on device', () => expect(file).toMatchSnapshot());

  it('builds without diagnostic messages', () =>
    expect(mockDiagnosticHandler).not.toBeCalled());

  it('emits an error if targeting SDK version lower than 3.0', async () => {
    mockSDKVersion.mockReturnValue({ major: 2, minor: 0 });
    await expect(compileFile('i18n.js')).rejects.toThrowError();
    expect(mockDiagnosticHandler.mock.calls[0]).toMatchSnapshot();
  });
});

it('emits sourcemaps with source paths relative to the project root', async () => {
  const file = await compileFile('sourcemap/index.js', { outputDir: 'app' });
  expect(file.sourceMap.sources).toMatchSnapshot();
});

it('applies specified polyfills', () => {
  return expect(
    compileFile('polyfill.js', {
      polyfills: {
        cbor: "export * from 'cbor'; export const encode = false;",
      },
    }).then(getVinylContents),
  ).resolves.toMatchSnapshot();
});
