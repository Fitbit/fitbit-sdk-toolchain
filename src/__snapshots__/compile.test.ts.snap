// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`allows importing image files when building settings 1`] = `
"\\"use strict\\"
var A=\\"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAIAAAACCAYAAABytg0kAAAKs2lDQ1BJQ0MgUHJvZmlsZQAASImVlwdUE+kWx+9MeoNAQgSkhN4EKdKl1wAK0sFGSCihxJAQVOzK4gquBRURLCu6CqJgo4gNsWBbFHtfkEVAfS6uYkPNG+AR3r533nvn/c+58/3OnW/u3PlmvnP+A0B/wJdIslAmQLY4VxoZ5MuNT0jkkrqACCRggTEw+QKZxCciIgwwjY1/1Yd7gAyPt62Ha/37+f8qNWGKTACARGCcLJQJsjE+jkWLQCLNBcBhAUbzcyXDXIYxW4o1iPHBYU4b5ZZhTh7lOyNzoiP9MO4DINP5fGkaAO09lufmCdKwOnQ2xrZioUiMsT/GnoJ0vhDjAownZWfPG+bDGJsn/1OdtL/UTFbW5PPTlDz6LCMi+4tkkiz+wv9zOf63srPkY/cwxIKeLg2OxEZNbM2qMueFKlmcPD18jEXCkfkjnC4PjhljgcwvcYyFfP/QMZZnxviMMV86fq0olxc9xtJ5kcr64qzpYcr6KTwlp8gCosY4VRTIG+P89Oi4Mc4TxU4fY1lmVOj4HD9lXiqPVPacKg1UPmO2bLw3AX/8Xrnp0cHjPcQr+xGm+Aco8+IY5XxJrq+ypiQrYrz/rCBlXpYXpbw2F/vAxjiDHxIxXidCuT4QBkHAhRjIglyQAh8CQQRiSMlNWTD8TYPfPMlCqSgtPZfrg+2aFC5PLLCZxLW3tXMGGN6Do6/4HWdkbyGcq+O5VdUAHicUCsXJ8VzITYAjSQDU+vGc+WwAZi/A5VMCuTRvNIcfPhCACqrABi3QAyMwB2uwBydwB28IgBAIh2hIgDkggHTIxjqfD4thBRRCMWyALVAOu2APVMEhOAqNcArOwSW4BjfhLjyGTuiBVzAAH2AIQRASwkBYiBaij5ggVog94oJ4IgFIGBKJJCBJSBoiRuTIYmQVUoyUIOXIbqQaOYKcQM4hV5AO5CHShfQjfyJfUBxKR9moLmqKTkZdUB80FI1GZ6NpaA6ajxag69AytBI9iDag59Br6F20E32FDuIAR8NxcAY4a5wLzg8XjkvEpeKkuKW4IlwprhJXi2vGteFu4zpxr3Gf8UQ8C8/FW+Pd8cH4GLwAn4Nfil+LL8dX4RvwF/C38V34Afx3AoOgQ7AiuBF4hHhCGmE+oZBQSthHqCdcJNwl9BA+EIlEDtGM6EwMJiYQM4iLiGuJO4h1xBZiB7GbOEgikbRIViQPUjiJT8olFZK2kQ6SzpJukXpIn8g0sj7ZnhxITiSLySvJpeQD5DPkW+Re8hCFSTGhuFHCKULKQsp6yl5KM+UGpYcyRFWjmlE9qNHUDOoKahm1lnqR+oT6jkajGdJcaTNoItpyWhntMO0yrYv2ma5Ot6T70WfR5fR19P30FvpD+jsGg2HK8GYkMnIZ6xjVjPOMZ4xPKiwVGxWeilBlmUqFSoPKLZU3qhRVE1Uf1Tmq+aqlqsdUb6i+ZlKYpkw/Jp+5lFnBPMG8zxxUY6nZqYWrZautVTugdkWtT52kbqoeoC5UL1Dfo35evZuFYxmx/FgC1irWXtZFVg+byDZj89gZ7GL2IXY7e0BDXWOKRqzGAo0KjdManRwcx5TD42Rx1nOOcu5xvkzQneAzIWXCmgm1E25N+Kg5UdNbM0WzSLNO867mFy2uVoBWptZGrUatp9p4bUvtGdrztXdqX9R+PZE90X2iYGLRxKMTH+mgOpY6kTqLdPboXNcZ1NXTDdKV6G7TPa/7Wo+j562XobdZ74xevz5L31NfpL9Z/6z+S64G14ebxS3jXuAOGOgYBBvIDXYbtBsMGZoZxhiuNKwzfGpENXIxSjXabNRqNGCsbzzNeLFxjfEjE4qJi0m6yVaTNpOPpmamcaarTRtN+8w0zXhm+WY1Zk/MGeZe5jnmleZ3LIgWLhaZFjssblqilo6W6ZYVljesUCsnK5HVDquOSYRJrpPEkyon3bemW/tY51nXWHfZcGzCbFbaNNq8mWw8OXHyxsltk7/bOtpm2e61fWynbhdit9Ku2e5Pe0t7gX2F/R0HhkOgwzKHJoe3U6ympEzZOeWBI8txmuNqx1bHb07OTlKnWqd+Z2PnJOftzvdd2C4RLmtdLrsSXH1dl7mecv3s5uSW63bU7Q93a/dM9wPufVPNpqZM3Tu128PQg++x26PTk+uZ5PmzZ6eXgRffq9LrubeRt9B7n3evj4VPhs9Bnze+tr5S33rfj35ufkv8Wvxx/kH+Rf7tAeoBMQHlAc8CDQPTAmsCB4IcgxYFtQQTgkODNwbf5+nyBLxq3kCIc8iSkAuh9NCo0PLQ52GWYdKw5mnotJBpm6Y9mW4yXTy9MRzCeeGbwp9GmEXkRJycQZwRMaNixotIu8jFkW1RrKi5UQeiPkT7Rq+PfhxjHiOPaY1VjZ0VWx37Mc4/riSuM35y/JL4awnaCaKEpkRSYmzivsTBmQEzt8zsmeU4q3DWvdlmsxfMvjJHe07WnNNzVefy5x5LIiTFJR1I+soP51fyB5N5yduTBwR+gq2CV0Jv4WZhf4pHSklKb6pHaklqX5pH2qa0/nSv9NL01yI/UbnobUZwxq6Mj5nhmfszFVlxWXXZ5Oyk7BNidXGm+MI8vXkL5nVIrCSFks4ct5wtOQPSUOk+GSKbLWvKZWNm57rcXP6DvCvPM68i79P82PnHFqgtEC+4vtBy4ZqFvfmB+b8swi8SLGpdbLB4xeKuJT5Ldi9FliYvbV1mtKxgWc/yoOVVK6grMlf8utJ2ZcnK96viVjUX6BYsL+j+IeiHmkKVQmnh/dXuq3f9iP9R9GP7Goc129Z8LxIWXS22LS4t/rpWsPbqT3Y/lf2kWJe6rn290/qdG4gbxBvubfTaWFWiVpJf0r1p2qaGzdzNRZvfb5m75UrplNJdW6lb5Vs7y8LKmrYZb9uw7Wt5evndCt+Kuu0629ds/7hDuOPWTu+dtbt0dxXv+vKz6OcHu4N2N1SaVpbuIe7J2/Nib+zetl9cfqnep72veN+3/eL9nVWRVReqnaurD+gcWF+D1shr+g/OOnjzkP+hplrr2t11nLriw3BYfvjlkaQj946GHm095nKs9rjJ8e31rPqiBqRhYcNAY3pjZ1NCU8eJkBOtze7N9SdtTu4/ZXCq4rTG6fVnqGcKzijO5p8dbJG0vD6Xdq67dW7r4/Px5+9cmHGh/WLoxcuXAi+db/NpO3vZ4/KpK25XTlx1udp4zelaw3XH6/W/Ov5a3+7U3nDD+UbTTdebzR1TO87c8rp17rb/7Ut3eHeu3Z1+t+NezL0H92fd73wgfND3MOvh20d5j4YeL39CeFL0lPm09JnOs8rfLH6r63TqPN3l33X9edTzx92C7le/y37/2lPwgvGitFe/t7rPvu9Uf2D/zZczX/a8krwael34N7W/bX9j/ub4H95/XB+IH+h5K32r+HPtO613+99Ped86GDH47EP2h6GPRZ+0PlV9dvnc9iXuS+/Q/K+kr2XfLL41fw/9/kSRrVBI+FL+iBXAYYGmpgL8uR+AkQDAwjwEVWXUI48IGfX1IwT/iUd99IicAGqxYdga+bUAHMbCdDmAqjfAsEWK9gbUwUEZ/5As1cF+tBYdc5qETwrFO10AUjPAN6lCMbRDofi2F2v2IUBLzqg3H5Ye9p8wUw1wRrLbzMHl8C/6O/OZCsdds9CvAAAACXBIWXMAABYlAAAWJQFJUiTwAAACAGlUWHRYTUw6Y29tLmFkb2JlLnhtcAAAAAAAPHg6eG1wbWV0YSB4bWxuczp4PSJhZG9iZTpuczptZXRhLyIgeDp4bXB0az0iWE1QIENvcmUgNS40LjAiPgogICA8cmRmOlJERiB4bWxuczpyZGY9Imh0dHA6Ly93d3cudzMub3JnLzE5OTkvMDIvMjItcmRmLXN5bnRheC1ucyMiPgogICAgICA8cmRmOkRlc2NyaXB0aW9uIHJkZjphYm91dD0iIgogICAgICAgICAgICB4bWxuczpleGlmPSJodHRwOi8vbnMuYWRvYmUuY29tL2V4aWYvMS4wLyIKICAgICAgICAgICAgeG1sbnM6dGlmZj0iaHR0cDovL25zLmFkb2JlLmNvbS90aWZmLzEuMC8iPgogICAgICAgICA8ZXhpZjpQaXhlbFlEaW1lbnNpb24+ODwvZXhpZjpQaXhlbFlEaW1lbnNpb24+CiAgICAgICAgIDxleGlmOlBpeGVsWERpbWVuc2lvbj44PC9leGlmOlBpeGVsWERpbWVuc2lvbj4KICAgICAgICAgPHRpZmY6T3JpZW50YXRpb24+MTwvdGlmZjpPcmllbnRhdGlvbj4KICAgICAgPC9yZGY6RGVzY3JpcHRpb24+CiAgIDwvcmRmOlJERj4KPC94OnhtcG1ldGE+CnrnRo4AAAATSURBVAgdY2RgYPgPxAxMIAIEAA4oAQP/6uhrAAAAAElFTkSuQmCC\\"
console.log(A)
"
`;

exports[`build fails when a non-existent relative import is specified 1`] = `"Could not resolve './__doesNotExist.js' from src/__test__/compile/relativeImportNotFound.js"`;

exports[`build fails when an absolute import is specified 1`] = `/foo.js is imported by <cwd>/src/__test__/compile/absoluteImport.js, but absolute imports are disallowed`;

exports[`build fails when an unrecognized binary file is imported 1`] = `Failed to compile <cwd>/src/__test__/compile/randomData.js`;

exports[`build fails when unknown external imports are used 1`] = `"Compile failed."`;

exports[`build passes when building a minimal app 1`] = `
"\\"use strict\\"
console.log(\\"hello world!\\")
"
`;

exports[`build passes when compiles a JS file with tslib helpers imported 1`] = `
"\\"use strict\\"
var o=function(t,n){o=Object.setPrototypeOf||{__proto__:[]}instanceof Array&&function(o,t){o.__proto__=t}||function(o,t){for(var n in t)if(t.hasOwnProperty(n))o[n]=t[n]}
return o(t,n)}
function t(t,n){o(t,n)
function r(){this.constructor=t}t.prototype=n===null?Object.create(n):(r.prototype=n.prototype,new r)}var n=function(){function o(){}o.prototype.doFoo=function(){}
return o}()
var r=function(o){t(n,o)
function n(){return o!==null&&o.apply(this,arguments)||this}n.prototype.doBar=function(){console.log(\\"foo\\")}
return n}(n)
var e=new r
e.doBar()
"
`;

exports[`build passes when compiles a JS file with tslib helpers inline 1`] = `
"\\"use strict\\"
var t=undefined&&undefined.__extends||function(){var t=function(n,o){t=Object.setPrototypeOf||{__proto__:[]}instanceof Array&&function(t,n){t.__proto__=n}||function(t,n){for(var o in n){if(n.hasOwnProperty(o))t[o]=n[o]}}
return t(n,o)}
return function(n,o){t(n,o)
function r(){this.constructor=n}n.prototype=o===null?Object.create(o):(r.prototype=o.prototype,new r)}}()
var n=function(){function t(){}t.prototype.doFoo=function(){}
return t}()
var o=function(n){t(o,n)
function o(){return n!==null&&n.apply(this,arguments)||this}o.prototype.doBar=function(){console.log(\\"foo\\")}
return o}(n)
var r=new o
r.doBar()
"
`;

exports[`build passes when known external imports are used 1`] = `
"\\"use strict\\"
function e(e){return e&&typeof e===\\"object\\"&&\\"default\\"in e?e[\\"default\\"]:e}var t=e(require(\\"cbor\\"))
t()
"
`;

exports[`build passes when require calls are unmodified 1`] = `
"\\"use strict\\"
var r=require(\\"cbor\\")
r()
var e=require(\\"./foo.js\\")
e()
"
`;

exports[`build passes when tslib when imported 1`] = `
"\\"use strict\\"
function t(i){return this instanceof t?(this.v=i,this):new t(i)}t()
"
`;

exports[`emits diagnostics given JS code with bad syntax 1`] = `
Array [
  Object {
    "category": 1,
    "file": Object {
      "path": <cwd>/src/__test__/compile/badSyntax.js,
      "position": Object {
        "end": Object {
          "character": 17,
          "line": 0,
        },
        "start": Object {
          "character": 17,
          "line": 0,
        },
      },
    },
    "messageText": "TS1002: Unterminated string literal.",
  },
]
`;

exports[`emits diagnostics given TS code with a type error 1`] = `
Array [
  Object {
    "category": 1,
    "file": Object {
      "path": <cwd>/src/__test__/compile/typeError.ts,
      "position": Object {
        "end": Object {
          "character": 5,
          "line": 0,
        },
        "start": Object {
          "character": 4,
          "line": 0,
        },
      },
    },
    "messageText": "TS2322: Type '1' is not assignable to type 'string'.",
  },
]
`;

exports[`when allowUnknownExternals is enabled successfully builds an app with unknown external imports 1`] = `
"\\"use strict\\"
function e(e){return e&&typeof e===\\"object\\"&&\\"default\\"in e?e[\\"default\\"]:e}var t=e(require(\\"_never_exists_\\"))
t()
"
`;

exports[`when allowUnknownExternals is enabled successfully builds an app with unknown external imports 2`] = `
Array [
  Array [
    Object {
      "category": 0,
      "messageText": "_never_exists_ is imported by src/__test__/compile/unknownExternalImport.js, but could not be resolved",
    },
  ],
]
`;

exports[`when building a device component which uses gettext polyfills gettext on device 1`] = `
"\\"use strict\\"
var e=require(\\"resources\\")
function r(r){var t=e.getResource(\\"text\\",\\"_\\"+r)
if(t===null)return r
return t}console.log(r(\\"hello\\"))
"
`;

exports[`when compiling a module with statements above an import declaration builds 1`] = `
"\\"use strict\\"
require(\\"util\\")
var e=require(\\"cbor\\")
console.log(\\"Hello!\\")
"
`;

exports[`when targeting SDK 1.0 allows JSON imports 1`] = `
"\\"use strict\\"
var t=\\"fitbit\\"
console.log(t)
"
`;

exports[`when targeting SDK 1.0 allows unintentionally non-relative imports 1`] = `
"\\"use strict\\"
function n(){return 1}console.log(n())
"
`;

exports[`when targeting SDK 1.0 allows unintentionally non-relative imports 2`] = `
Array [
  Object {
    "category": 0,
    "messageText": "Import in src/__test__/compile/incorrectlyNonRelativeImport.js has non-relative path but should be relative: replace 'importee' with './importee'.",
  },
]
`;

exports[`when targeting SDK 2.0 does not allow JSON imports 1`] = `
<cwd>/src/__test__/compile/data.json: Unexpected token, expected ";" (2:7)

[0m [90m 1 | [39m{[0m
[0m[31m[1m>[22m[39m[90m 2 | [39m  [32m"foo"[39m[33m:[39m [32m"bar"[39m[33m,[39m[0m
[0m [90m   | [39m       [31m[1m^[22m[39m[0m
[0m [90m 3 | [39m  [32m"pebble"[39m[33m:[39m [32m"fitbit"[39m[0m
[0m [90m 4 | [39m}[0m
[0m [90m 5 | [39m[0m
`;

exports[`when targeting SDK 3.0 emits ES5 code for device 1`] = `
"\\"use strict\\"
var o=function(){return\\"foo\\"}
console.log(o())
"
`;

exports[`when targeting SDK 3.0 emits ES6 code 1`] = `
"\\"use strict\\"
const o=()=>\\"foo\\"
console.log(o())
"
`;
